### ISSUE TITLE
NodeCG Core Enhancement Roadmap

### ISSUE THREAD

**Issue by Hoishin** [nodecg/nodecg#565]
I am aware that NodeCG has a lot of problem for developer and users.
The aim is to adopt to more modern technology, architecture and libraries, while keeping breaking change as less as possible.

Currently this is just a list of what I can think of to improve experience with NodeCG. I will be updating as we discuss more and polish ideas.

Any suggestion is welcome!

## TypeScript
- 2.0 will be fully refactored to TypeScript

## New client side framework (something not Polymer)
- Because Polymer is deprecated
- React, Preact, LitElement or Vue

## Bundle project structure
- Let you install NodeCG in your bundle, rather than installing bundles in NodeCG repo
- Will dramatically reduce steps to set up or run bundles.
- [I've already made a library to do this](https://github.com/Hoishin/include-nodecg) and mostly works, but there are some features that doesn't work or bugged (notably assets).
- Doing this and keeping the current structure shouldn't be hard, and we should do it.
- Also should support multiple bundles somehow.

## Make this repository monorepo
- Better way to share code and can easily publish part of code
- `@nodecg/dashboard`, `@nodecg/graphics` to be used from respective client sides, and `@nodecg/extension` to be used from extension
- Again, doing this and keeping the current "installing bundles in NodeCG repo" structure shouldn't be hard, and we should do it.

## Split client-side code and distribute separately
- It has been requested if it's possible to put dashboard/graphics assets somewhere else
- Should be doable without breaking change. Monorepo (above) should help.

## Bundle managing GUI
- Users shouldn't have to open console and run commands to run NodeCG
- Some kind of Electron app that can pull and run bundles, and also provides bundle's dashboard UI instead of browsers.
- [There was something like this](https://github.com/nodecg/dashboard), but with totally different context and purpose. Could be used as a starting point.

**Comment by Dan-Shields** [comment]
> New client side framework

Depending on what time frame we're looking at for this upgrade, I'd pitch Vue 3, with something like Buefy/Ionic for UI. It's component format its roughly similar to Polymer's so could make porting *fairly* straightforward? (I also personally prefer developing with it over react).

> Bundle project structure

Lange discussed this quite a bit in #395, but the biggest problem with all the solutions I see right now is multiple-bundle support. As Lange discovered in #138, for helper bundles, there is a better solution he discovered with [`nodecg-obs`](https://github.com/nodecg/nodecg-obs), which is to make the bundle developer include the extension modules in their own extensions, and include dashboard panels as widgets in their own panels.

This "mixin" style of inheritance could be applied to other helper bundles I'm sure, and could potentially eliminate the need for multiple-bundle setups all together. That being said, I very rarely use more than one bundle, so maybe there are use cases that aren't suited to this I'm missing?

> Bundle managing GUI

Sounds like a great idea. It could potentially come with an admin panel for changing user permissions/basic config settings.

**Comment by EwanLyon** [comment]
User access level's should be on the plan as well. (e.g. Volunteer and Admin levels)

**Comment by TBSliver** [comment]
> New Client Side Framework

Im assuming this is just for the core NodeCG Client and will not impact bundle developers at all? I remember getting massively confused in the beginning and developing my entire bundle in Polymer originally due to that :(

Note with Vue vs. React, the biggest question is probably more how comfortable are the current core developers, and how easily do you want others to be able to be able to contribute to core. Obviously both have their own advantages and drawbacks, but something to keep in mind as well!

> Panel Management

I would love for the panel management to no longer be fixed by the bundle - the easiest way I could see without much change would be to just track workspace panel assignment in a Replicant as I guess the order of the panels is stored client side.

> Bundle Project Structure

I would love to see NodeCG installed as a dependency instead of running from a checkout, would also remove the requirement for installing Git for basic installation.

A big question here, is how do people structure their bundles? I pivot between multiple small bundles that talk together and one large bundle that does everything depending on what it is im working on - but definitely having the ability to have small seperate bundles would make certain things easier for new users. As an example, Im steadily moving all my Twitch API integration into its own bundle so that its all in one place, then just use messages and replicants from there. Combined with the ability to move the 'login' panel to wherever you want on another dashboard would be massively useful (for the panel management gui)

As for the actual structure, I would look towards NextJS for a rough idea, where they only work with a particular folder structure - but maybe slightly more flexible as needed:

* Root Folder
  * `node_modules`
  * `graphics` - graphics for 'parent' bundle
  * `dashboards`  - dashboards for 'parent' bundle
  * `extensions` - extensions for 'parent' bundle
  * `config` - config storage folder
  * `bundles` - works the same as current bundles folder
  * `.nodecg` - Move all the nodecg managed items here

To make it flexible, could add some more variables under a NodeCG key in package.json:

```json
{
  "nodecg" : {
    "graphics_dir": "./graphics",
    "dashboards_dir": "./bundles",
    "extensions_dir": "./extensions",
    "config_dir": "./config",
    "bundles_dir": "./bundles"
  }
}
```

Also a way to track bundles installed could be done under that key or somewhere, using the same arguments passed to nodecg-cli for bundle installation (and possibly version/branch? depending on bundle storage method (git vs zip etc)).

Using a dot dir for storing all the NodeCG specific stuff means its out of the way, and gives you cache building directories without polluting any bundles main namespace.

Note the package.json could easily be separated down further to have `nodecg.graphics.directory` to allow for further configuration of graphics specifically. That whole section could also easily be put into a `nodecg.json` file as well for top level config instead of package.json, which could allow for fully custom ci builds etc. - I would love to be able to add a build step that means I don't have to think about adding more dashboards to package.json automatically for example (and not break my package.json in the process).

> Bundle Packages

Having the graphics/extensions/dashboards modules separate would be great as well in the above structure as it would reduce the amount of duplicated node_modules parts (obviously keeping it that each bundle manages its own dependencies etc.)

> Bundle Management GUI

Something that would give you the ability to download and run, would be great. If it could prompt to create/open a folder to run as a bundle manager and interface that would be awesome as well. Especially if it could 'restart' the NodeCG server seperately to itself, so it was more like a nodecg-cli but a GUI :D

As you already allow for json schemas for the config, I think there are some automatic UI's that read those to make interfaces, so that would fork for bundle config - with fall-back to a basic text editor ([ace editor](https://ace.c9.io/) for example) when there isn't a json schema available.

For downloading/installing bundles, if it could check for the existence of git and then if its not there, download bundles using normal 'release' downloads, that would be great for simplified installs - allowing for just this app and NodeJS to be installed. Another option would be to look at fall-back to some git implementations in JS, of which there is several ([isomorphic-git](https://github.com/isomorphic-git/isomorphic-git) springs to mind)

Note it would have to have buttons for npm update, and probably the ability to run other commands in package.json for each bundle such as building if they're using webpack of parcel or some such to save people from having to use the terminal!

But yes, something like this would be quite cool - I definitely already have a need for this as it would allow me to help some friends who arent technically minded, but are wanting some of the things that NodeCG can do that Streamelements and Streamlabs cant manage for example.

----

I talk too much...

**Comment by grantjbutler** [comment]
Really nice to see an issue here talking about, at a high level, the kinds of changes to make to NodeCG and its overall evolution as a project. I don't have many thoughts on most of these things, but I have written out a few things below, including a few additional topics:

### Bundle project structure

Multi-bundle support is a must for me. My biggest use case for NodeCG is to create overlays for charity gaming marathons, and the all have a lot of the same requirements: fetch donations from some service, display donations on screen, display the current total, have some kind of ticker with additional information, have the ability to set donation goals and other incentives. Since they all need the same general behavior for these functions, but differ in how that information is displayed in the overlay, I've taken the approach of "renderless bundles," for lack of a better term. That is, bundles that only provide logic and replicants in the form of an extension, and some kind of management UI in the form of dashboard panels. No actual graphics are included with these bundles, since the actual display of this information differs from event to event. But they all need the same underlying functionality (the same structure of data, and the same dashboard panels to manage the data), so that part can be re-used. I get that this is not how everyone creates their bundles, but it is an approach that has worked well for me, and it would be nice to see this use case supported with whatever new approach for installing bundles we come up with.

### Bundle managing GUI

Huge +1 for this. This sounds like a great way to simplify the managing of a NodeCG installation, especially in the use case of a developer being contracted to build a bundle for some event. With such GUI like this, it might be possible for the developer to set everything up before hand, zip up the installation, send it to the event, and just say "run this app and you're good to go."

Electron probably makes the most sense for a cross-platform app. I think trying to use some other cross-platform technology (like C++ and Qt, for example) wouldn't be the right approach, as it requires knowledge of those other languages and libraries. We should continue to build on the technologies that we know to make it easier to manage and develop this app.

As @TBSliver points out, this application would probably be a good place for providing some GUI-based config editor for bundles (we've got #487, so this is already an area we think needs improvement). This application might also be a good place for the replicant explorer (#484) to live as well.

I wonder if this and the proposed changes to bundle structure would go hand in hand? That is, separate the API of NodeCG and the running of a NodeCG instance. If I'm developing a bundle, and I `npm install @nodecg/extension` or something, that doesn't install all the things needed to run an instance of NodeCG. Rather, it just gives me a library that has the interface and APIs needed for extensions. To actually run a NodeCG instance, I need the electron app, or the cli tool.

This means I'm still "installing" my bundle into NodeCG, but that's only for when I want to actually run the bundle. That is, I'm not required to install the bundle into NodeCG if I'm just doing development. That said, this might be making things more complex for the development process, as it's now a few more steps if I want to test something. But on the flip side, maybe that's okay? That is, developers would basically have a single install of NodeCG that would pull in all the bundles they're working on, and then that's canonical installation of NodeCG that can more or less always be running. I'm not sure how other people have their dev environment set up, so maybe this isn't really worth it? ¯\\\_(ツ)\_/¯

### Other Points

#### Workspace Improvements

I feel like, with some of the things in the pipeline (namely the user auth/permission changes), as well as some limitations that exist now, there should be improvements made to how workspaces are created and managed. My gut is that workspaces should change from being things that are defined by the bundle's dashboard panels, to being something created by the users of NodeCG. By doing this, there is the possibility of creating workspaces specifically tailored to a specific user types/permission levels, ensuring they have only the dashboard panels they need to do their job, and no access to anything else.

Additionally, by making this change to how workspaces are defined, we open up the possibility of the same panel being in multiple workspaces, something that's not possible now. This allows multiple user types/permission levels to have access to the same controls, but allows for different layouts depending on the needs of that user type or permission level. We could still have something implemented where bundle developers could provide "workspace templates," if you will, that would be a pre-defined collection of dashboard panels that can be just dropped in to NodeCG as a pre-made workspace.

#### Media/Assets

My gut tells me there are some improvements that can be made asset management, as well as the APIs for interacting with assets. Unfortunately, my experience with with this part of NodeCG is somewhat limited, just because I haven't gotten around to building things that need that functionality yet. Most of this is also going off of what I vaguely remember from conversations in the previous Discord server, so I sadly don't have any specific feedback on this topic, just a vague "feels like this feature could use some attention" feelings.

**Comment by grantjbutler** [comment]
> I propose each NodeCG project needs to provide an index.js file which includes the NodeCG app, loads the core bundles, then starts the server

I don't think this does anything that the GUI app wouldn't do (and, by extension, the CLI tool with an appropriate config file). My gut says that you wouldn't want to do any kind of custom server mounting inside of this `index.js` file you're proposing, because any of that mounting is specific to that project (and thus not necessarily portable). I might not be able to take a bundle I have and move it to another project and have it just work without also having to copy all or part of this `index.js` file around. @Dan-Shields what kinds of further extension to the server would you expect to be done in this `index.js` file?

Additionally, I worry that this is going to become boilerplate-y. That is, if I'm doing development on a bundle, this file will essentially look something like this:

```js
const nodecg = require('@nodecg/server');
nodecg.load('.');

const server = nodecg.server(9090, 'localhost');
server.start();
```

I fear I'm going to have this same file for every bundle I'm working on, with little to no modification between bundles.

> Note that this index.js file would only be used if the bundle was run directly; when a bundle is loaded as a dependency, it's ignored.

This is a little confusing to me. What is "run directly" in this context? Is this if I have a bundle with an `index.js` file in the root directory of the bundle, my current working directory is the root of the bundle, and then I run some kind of command that starts the server based off of the configuration in the `index.js` file?

If so, what happens if I have multiple bundles? Do I need to create a new folder that contains all of these bundles, and thus a new `index.js` file, making sure to copy in relevant parts from the bundle's `index.js` file?

This is super heckin' rough, but this is basically what I imagined the GUI tool would look like:

<img width="581" alt="Screen Shot 2020-10-08 at 12 21 27 PM" src="https://user-images.githubusercontent.com/526054/95486289-d3366580-0960-11eb-96af-5ac09e2fc493.png">

I have an install of NodeCG (this may be represented as an actual folder on disk that contains these bundles, or this is configuration information about an install is persisted into some kind of "user settings" folder or something, it's not really important for the purposes of this). I can add bundles to this install, I would imagine more or less as they exist today. There is API exposed to allow for mounting of endpoints and routes to the server, but it's not the responsibility of the bundles to worry about running the server or other aspects of NodeCG; that responsibility is the GUI app's.

We can build a whole load of other tolling into this GUI app (I've got a GUI for editing a bundle's config, but the replicant explorer could go here, as well as a browser for opening workspaces as their own windows, separate from a dedicated web browser), but to run NodeCG, you'll need this app (or the CLI tool. I'm imagining that everything you could do in this GUI to run NodeCG, you could do that with the CLI tool and a config file. In fact, maybe the portion of managing the install and running the server in the GUI app is just a wrapper around managing the config file and invoking the CLI tool).

Okay, I feel like I'm starting to ramble. I hope this kind of brain dump makes sense. Please ask questions if something doesn't.

**Comment by Dan-Shields** [comment]
> what kinds of further extension to the server would you expect to be done in this index.js file?

To quote(ish) Lange, while NodeCG remains a program you run, and not a library you import "it becomes very hard or impossible for a bundle author to extend or override any of NodeCG's behaviors and systems (for instance, to add support for a single-sign-on authentication, or whatever the hot new dev server with hot module reloading is, etc)". 

If NodeCG is a library, people can import whatever other libraries they want and have access to the core NodeCG app, rather than needing to rely on `.mount` in extensions. A plugin-type API could be developed to accommodate SSO/build tool libraries as core NodeCG mods.

Other applications could be custom permissions systems, better multi-user support etc. Stuff that might not necessarily be in core NodeCG, but important for specific use-cases.

> I worry that this is going to become boilerplate-y

Yeah rethinking it, maybe it would be better to define the bundle paths in a `.nodecgrc` file or similar, in place of all the `.load` calls I talked about. That way, the GUI could read this as well, and if you wanted to start from the command line, you could do:

```javascript
const nodecg = require('@nodecg/server')('./.nodecgrc');
nodecg.start();
```

And to run the server, you use `node .`. No CLI program is required.

Perhaps `index.js` isn't the right term. Maybe `launch.js` is better (though really it can be anything). If a bundle has one of these files, it's saying "I can be run as the root bundle, run this to set me up". I'm envisaging you have one `launch.js` file for each event you run, with it usually looking like the one above. It isn't a requirement to develop a bundle, only for when you want to run NodeCG from the command line.

This adds a small extra step for setting up an entirely new project using the command line, but opens the door for more creative uses for NodeCG. As Lange said, someone might want to integrate an SSO portal into their setup, and `launch.js` provides a place for that. 

> I don't think this does anything that the GUI app wouldn't do (and, by extension, the CLI tool with an appropriate config file)

This isn't an alternative to the GUI, just a change to how NodeCG is run from the command line.

Lastly, an important point to make is that in my proposal, all bundle dependencies could be npm modules (or at least installed through npm from a git repo) and so are installed in the parent bundle with `npm install`. This removes the need to install bundles with a custom CLI or GUI tool, which is a practice that seems clunky to me when we have so much infrastructure for installing dependencies with node already. The proposed GUI would still be a very useful config and server start-up tool.

### TL;DR
I propose:
- NodeCG becomes an npm library (rather than a program) that bundles can depend on, and you pick a bundle to run as the "parent bundle" or "root project".
- Developers use npm to install bundle dependencies (or manually copy them) to the parent bundle, defining the include paths in `.nodecgrc`.
- To run a parent bundle from a command line, a JavaScript file is needed with a small amount of boilerplate.
- Alternatively, the parent bundle can be run from the GUI.

**Comment by DomHynes** [comment]
In terms of concerns around boilerplate, there's probably some lessons to be learnt from frameworks like Next.js, which have sensible defaults and easy project starts, but support implementing a custom server. 

The nodecg server could expose a cli that picks up something like a .nodecgrc and runs the server with default settings, with something like `nodecg start`, or if you need to hook into the express server/provide more detailed configuration, run something like example above in js.

**Comment by Dan-Shields** [comment]
Sensible defaults are definitely a must, agreed. We want to keep the time & steps it takes to get an empty NodeCG instance up and running as low as possible.

That being said, I think using a CLI tool to run what could be as little as two lines of JS would be a bit overkill. Instead, perhaps `nodecg setup` could be reimagined as a [generator-nodecg](https://github.com/nodecg/generator-nodecg) replacement which creates the requisite empty directories, package.json (setting @nodecg/server as a dependency) and a sensible launch script, which is then bound to `npm start` or similar. If really desired, `nodecg start` could then just become an alias to `npm start`, or `node launch.js`.

I really think this increased transparency of just having a .js file that you run with node is a benefit to users trying to get to grips with how NodeCG works, as opposed to them them being told to install a CLI tool that abstracts everything away from you.

In my vision for this setup, the process for installing a ready-made, multi-bundle NodeCG setup looks like:

1. `git clone <repo URL>`
2. `npm i --production`
3. `npm start`

One repo clone, one npm install, no global npm packages, no having to manually install dependency bundles one-by-one.

**Comment by Dan-Shields** [comment]
I've got a bit of spare time right now and I'm keen to start working on a PoC for my above proposals, but first it'd be good to get your thoughts on the whole idea @Hoishin. As I mention somewhere in the above walls of text I realise it's a break from the backwards-compatibility-first ideal you were going for but I think the benefits are worth at least investigating.

**Comment by jweisbaum** [comment]
Hi all, I just found this library and am only just beginning to learn it. I am very impressed with the developer outreach in the docs and Discord. 

I am the Technical Director for a sports non-profit that's trying to build tools that pair geospatial tracking (like Strava but for weather and ocean nerds) with media streaming in order to grow various sports (sailing, windsurfing, etc.). My previous role was as a Developer Relations Engineer at Google. If I could just pitch adopting React instead of competing libraries without it devolving into a shouting match because I respect that the Vue devs out there know what they're talking about and I'm sure it's a kick ass framework.

I fully recognize from the purely technical perspective React may be either overkill or have no clear advantage. My axioms are the following: 1) This project has huge untapped potential, especially in the non-profit and startup world. 2) In order to reach that potential it must be "cool" and developer friendly. 3) No matter what frameworks are used, it's likely in 5-10 years (or less) it will have to be rebuilt or updated anyway. 

With those axioms in place let me just make some observations. React devs really love React. Yeah, maybe it distracts them sometimes from being objective, but I think it's a really good strategic choice to adopt tools that people love. The React community is absolutely huge. [This article](https://dzone.com/articles/react-vs-vue-in-2021-best-javascript-framework) has some stats on the relative size of the React vs Vue community, though who knows if it's accurate. Choosing a framework with a large and active community increases the odds that people will contribute to this project and all of the various React libraries will increase the types of bundles people create for it. There are numerous examples of amazing React libraries but [this one](https://github.com/keplergl/kepler.gl) stands out to me. 

Again, I don't really have a dog in this fight other than I want this project to take off so that live graphics are as accessible as live streaming and I've spent a long time as a professional (at a company who's SDKs often lose to Facebook's) thinking about how to make a community take off.

**Comment by c43721** [comment]
I can stand behind adopting React.

I love Vue and even Angular, but the React ecosystem is huge, and with the announcement of Next 11 and React 18, it's something to certainly consider.

I'm also a huge fan of the Nest framework, making backends with Express or Fastify super easy and has a wide range of packages to accomondate anything you need.

I really think that if there was a migration or even a hope to transfer to TS, you may as well just rewrite the entire project. I know this is a huge undertaking, but I really think that if it were to happen (i.e. the rewrite) then I think traction could move again.

I'm willing to put in effort too, I really like Node.cg and what it offers. I also think it still has standing in 2021. 

I know I brough this up before and was met with some hesitation, but I really really think this is an opportunity to help bring the codebase to more a more modern look and feel.

This also comes with the ability to use Node v14 syntax, as v12 is not LTS anymore. 

Although I may think it's all easy to rewrite and use these new tools, I really have no idea what I'm doing when it comes to building an application like this. But I think putting in ideas will only help contribute, and not halt.

**Comment by gtrabanco** [comment]
## Proposals

- Native development but allowing use of React or whatever you want when developing. Visually there are not a lot of stuff in nodecg there are more backend stuff. So, webcomponents could be an option.
- Server Sent Events instead of websockets. With api rest for comunication from client if necessary (commonly none) is enough. Graphics clients only need to read events and graphically show it with css/html/js.

**Comment by Dan-Shields** [comment]
```
comunication from client if necessary (commonly none)
```

I don't agree with this, I rely heavily on bidirectional websocket messages. How do Replicants fit into an SSE setup?

**Comment by gtrabanco** [comment]
> ```
> comunication from client if necessary (commonly none)
> ```
> 
> I don't agree with this, I rely heavily on bidirectional websocket messages. How do Replicants fit into an SSE setup?

Does not fit. But could use REST API which can be transparent and make possible the usage of unauthenticated graphics.

**Comment by duncte123** [comment]
> Some kind of Electron app that can pull and run bundles, and also provides bundle's dashboard UI instead of browsers.

Why electron? I get that it's "easier" on end users, but at the end of the day you're still running in a browser. I'd prefer this to be just a page on the browser dashboard.

**Comment by jhobz** [comment]
Wanted to throw out there that it'd be a very nice enhancement if the replicants database could be configured to pull from an existing database elsewhere, rather than using one that NodeCG places on the local filesystem. This is one of the few steps preventing NodeCG from being used in some serverless app platforms, where write access is restricted.

**Comment by Dan-Shields** [comment]
Didn't mean to unpin (misclicked on mobile) and now can't work how to undo on mobile.

Does raise the question if this issue is still relevant or should be renamed "ideas for a nodecg fork"? Don't see many (if any) of the suggestions here ever making it into this project in my opinion.
